tests/webview/integration/test_rex_redirects5.py

```diff
@@ -0,0 +1,50 @@
+import backoff
+import requests
+
+from tests import markers
+
+import csv
+import pytest
+
+
+@backoff.on_exception(backoff.expo, requests.exceptions.ConnectionError)
+def get_url(url):
+    return requests.get(url)
+
+
+@markers.slow
+@markers.nondestructive
+def test_calculus2_uri_redirect_to_rex(webview_base_url, rex_base_url, calculus_vol_2_uri):
+    # GIVEN a webview_base_url, rex_base_url and a calculus_vol_2_uri
+
+    # WHEN we go to a page based on the webview_base_url and uri
+    cnx_page_slug = calculus_vol_2_uri.split("/")[-1]
+    cnx_url = f"{webview_base_url}{calculus_vol_2_uri}"
+    response = get_url(cnx_url)
+
+    status_codes = []
+    error_urls = []
+
+    # THEN we are redirected to rex
+    if response.status_code != 200:
+
+        status_codes.append(str(response.status_code))
+        error_urls.append(cnx_url)
+
+        dict_urls = {"PAGE URL": error_urls[i] for i in range(len(status_codes))}
+        dict_errors = {"ERROR": status_codes[i] for i in range(len(status_codes))}
+        dict_urls_errors = {**dict_urls, **dict_errors}
+
+        with open("url_error_list.csv", "a") as urlerr:
+            writer = csv.DictWriter(urlerr, dict_urls_errors.keys())
+
+            if urlerr.tell() == 0:
+                writer.writeheader()
+
+            writer.writerow(dict_urls_errors)
+
+        pytest.fail(f"{response.status_code} in {cnx_page_slug}")
```

Comment:
As part of merging into `test_rex_redirects.py`, you can probably keep things DRY by pulling this out to a helper function like you did with `get_url`.

Reply:
@rnathuji I made the recommended changes but this one (helper function) I am not sure. Looking at it but don't see how...?

Reply:
🤔 - Maybe you could do something like:

```python
def log_failure_to_csv(status_code, cnx_url):
    dict_urls_errors = {"ERROR": status_code, "PAGE URL": cnx_url}

    with open("url_error_list.csv", "a") as urlerr:
        writer = csv.DictWriter(urlerr, dict_urls_errors.keys())
        if urlerr.tell() == 0:
            writer.writeheader()
        writer.writerow(dict_urls_errors)
```

...and then invoke as:

```python
    # THEN we are redirected to rex
    if response.status_code != 200:
        log_failure_to_csv(response.status_code, cnx_url)
        pytest.fail(f"{response.status_code} in {cnx_page_slug}")
    else:
        assert response.url.startswith(rex_base_url)
```

Or at least I think that will reduce some repetition when you move this test structure over to `test_rex_redirects.py`, but you can see what you think when doing that.

Reply:
hm... I was trying something similar but I was getting unknown parameters/variables... thanks for the tip and review :)

Reply:
@rnathuji thanks for the suggestion. Now I see where I went wrong when I tried the helper function. here are the changes. Also, I noticed that certain 404 errors will always occur because some urls lead to main chapters but they are not real pages. I am looking at the `ce-scripts/.venv_scripts/lib/python3.8/site-packages/rex_redirects.py` where all the url variants are created. If we could eliminate all fake pages (I think they start with #) in there, it would reduce many 404 failures and make the report more accurate.